File: 0322-coin-change/0322-coin-change.py
class Solution:

    def knapsack(self,coins,idx,amt,dp):
        if idx<0: 
            return float("inf")
        if amt==0:
            return 0 
        
        if dp[idx][amt] is not None:
            return dp[idx][amt]

        if coins[idx]>amt:
            dp[idx][amt]= self.knapsack(coins,idx-1,amt,dp)
            return dp[idx][amt]

        a=self.knapsack(coins,idx,amt-coins[idx],dp)
        # print(a)
        a = a if a==float("inf") else 1+a
        b = self.knapsack(coins,idx-1,amt,dp) 
        dp[idx][amt] = min(a,b)
        return dp[idx][amt]    

    def coinChange(self, coins: List[int], amount: int) -> int:
        coins.sort() 
        dp = [ [ None  for _ in range(amount+1)] for _ in range(len(coins))]
        ans =  self.knapsack(coins,len(coins)-1,amount,dp)
        return -1 if ans==float("inf") else ans

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinCombs(vector<int>& coins,int idx,int amt,vector<vector<int>>& dp){
        if(!amt) return 0;
        if(idx==0){
            if(amt%coins[idx]!=0) return 1e8;
            return amt/coins[idx];
        }
        if(dp[idx][amt]!=-1) return dp[idx][amt];

        if(amt>=coins[idx]){
            int take = 1+coinCombs(coins,idx,amt-coins[idx],dp);
            int notake = coinCombs(coins,idx-1,amt,dp);
            return dp[idx][amt]=min(take,notake);
        }
        return dp[idx][amt]=coinCombs(coins,idx-1,amt,dp);
    }

    
    int coinChange(vector<int>& coins, int amount) {
        // take or not take from high - low 
        sort(begin(coins),end(coins));
        int n = coins.size();
        vector<vector<int>> dp(n,vector<int>(amount+1,0));
        for(int i=1;i<=amount;i++){
            if(i%coins[0]) dp[0][i]=1e8;
            else dp[0][i]=i/coins[0];
        }
        for(int i=1;i<n;i++){
            for(int j=0;j<=amount;j++){
                if(j>=coins[i]){
                    int take = 1+dp[i][j-coins[i]];
                    int notake = dp[i-1][j];
                    dp[i][j]=min(take,notake);
                }
                else{
                    dp[i][j]=dp[i-1][j];
                }
            }
        }

        
        int ans =  dp[n-1][amount];
        return ans>=1e8?-1:ans;
    }


};